compiler/src/iree/compiler/Codegen/Common/CombineLayoutTransformation.cpp

```diff
@@ -0,0 +1,383 @@
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-combine-layout-transformation"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_COMBINELAYOUTTRANSFORMATIONPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+//===----------------------------------------------------------------------===//
+// Preprocessing Utilities
+//===----------------------------------------------------------------------===//
+
+/// Convert complex ops into simpler ops by decomposing or raising to a named
+/// op.
+///  - `PackOp`s and `UnPackOp`s are decomposed.
+///  - Transpose `linalg::GenericOp`s are raised to `linalg::TransposeOp`s.
+static void simplifyComplexRelayoutOps(RewriterBase &rewriter,
+                                       FunctionOpInterface funcOp) {
+  OpBuilder::InsertionGuard g(rewriter);
+  SmallVector<linalg::PackOp> packOps(
+      funcOp.getFunctionBody().getOps<linalg::PackOp>());
+  for (auto packOp : packOps) {
+    rewriter.setInsertionPoint(packOp);
+    (void)linalg::lowerPack(rewriter, packOp,
+                            /*lowerPadLikeWithInsertSlice=*/false);
+  }
+  SmallVector<linalg::UnPackOp> unPackOps(
+      funcOp.getFunctionBody().getOps<linalg::UnPackOp>());
+  for (auto unPackOp : unPackOps) {
+    rewriter.setInsertionPoint(unPackOp);
+    (void)linalg::lowerUnPack(rewriter, unPackOp,
+                              /*lowerUnpadLikeWithExtractSlice=*/false);
+  }
+  SmallVector<linalg::GenericOp> genericOps(
+      funcOp.getFunctionBody().getOps<linalg::GenericOp>());
+  for (auto genericOp : genericOps) {
+    if (linalg::isaTransposeOpInterface(genericOp)) {
+      rewriter.setInsertionPoint(genericOp);
+      (void)linalg::specializeGenericOp(rewriter, genericOp);
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Combining Layout Transformation Ops
+//===----------------------------------------------------------------------===//
+
+/// Folds an `op` that does not affect index computation into a `mapScatterOp`.
+/// This is used for ops like `linalg::CopyOp`.
+static IREE::LinalgExt::MapScatterOp
+foldIdentityLikeOpIntoMapScatter(RewriterBase &rewriter, Operation *op,
+                                 IREE::LinalgExt::MapScatterOp mapScatterOp) {
+  assert(mapScatterOp.getInput() == op->getResult(0) &&
+         "expected op to be the producer of mapScatterOp");
+  rewriter.modifyOpInPlace(
+      mapScatterOp, [&]() { mapScatterOp->setOperand(0, op->getOperand(0)); });
```

Comment:
nit: use `getInputMutable().assign(op->getOperand(0)))` when possible. It improves readability.

(I'm not sure if this one works or not, but other cases should work.)